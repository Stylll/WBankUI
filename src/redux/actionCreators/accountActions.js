import * as types from '../constants/actionTypes';

export const getAccounts = () => ({
  type: types.GET_ACCOUNTS,
});

export const getAccountsSuccess = (accounts) => ({
  type: types.GET_ACCOUNTS_SUCCESS,
  accounts,
});

export const getAccountsFailure = (error) => ({
  type: types.GET_ACCOUNTS_FAILURE,
  error,
});

export const createAccount = (accountData) => ({
  type: types.CREATE_ACCOUNT,
  accountData,
});

export const createAccountSuccess = (accountData) => ({
  type: types.CREATE_ACCOUNT_SUCCESS,
  accountData,
});

export const createAccountFailure = (error) => ({
  type: types.CREATE_ACCOUNT_FAILURE,
  error,
});
