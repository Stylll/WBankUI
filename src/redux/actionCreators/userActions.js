import * as types from '../constants/actionTypes';

export const signupUser = (userData) => ({
  type: types.SIGNUP_USER,
  userData,
});

export const signupUserSuccess = (userData) => ({
  type: types.SIGNUP_USER_SUCCESS,
  userData,
});

export const signupUserFailure = (error) => ({
  type: types.SIGNUP_USER_FAILURE,
  error,
});

export const logoutUser = () => ({
  type: types.LOG_OUT,
});
