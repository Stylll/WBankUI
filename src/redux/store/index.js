import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';
import axios from 'axios';

import rootReducer from '../reducers';
import rootSaga from '../sagas';
import { jwtKey } from '../../helpers/defaults';

const sagaMiddleware = createSagaMiddleware();
// const reduxRouterMiddleware = routerMiddleware(history);
const middleware = composeWithDevTools(applyMiddleware(sagaMiddleware));
const store = createStore(rootReducer, middleware);

// eslint-disable-next-line no-undef
const token = localStorage.getItem(jwtKey);

axios.defaults.headers.common.authorization = token;

sagaMiddleware.run(rootSaga);

export default store;
