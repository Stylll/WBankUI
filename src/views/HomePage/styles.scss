@import '../../partials/colors';

.account-container {
    display: flex;
    justify-content: center;

    .loader-container {
        margin-top: 50px;
        .message {
            font-size: 18px;
        }
    }

    table {
        border-collapse: collapse;
    }
      
    table, th, td {
        border: 1px solid $black;
    }

    th, td {
        padding: 15px;
        text-align: left;
    }

    tr:hover {
        background-color: $gray;
        cursor: pointer;
    }

    th {
        background-color: $primary-color;
        color: $white;
    }
}

.navigation-container {
    display: flex;
    margin: 20px;
    .navbox {
        padding: 10px;
        margin: 0 20px;
        box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
        color: $white;
        &:hover {
            cursor: pointer;
        }
    }
    .navbox.create {
        background-color: $green;
    }
    .navbox.deposit {
        background-color: $yellow;
    }
}

/* small to medium media query */
@media only screen and (max-width: 600px) {
    .account-container {
        justify-content: start;
        overflow-x: scroll;
        margin-left: 20px;
        margin-right: 20px;
    }
}
