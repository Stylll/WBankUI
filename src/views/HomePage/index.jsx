import React, {  useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Header from '../../components/Header/Header.component';
import Loader from '../../components/Loader/Loader.component';
import * as accountActions from '../../redux/actionCreators/accountActions';
import './styles.scss';

const HomePage = ({
  accounts, getAccounts, isLoading,
}) => {
  useEffect(() => {
    getAccounts();
  }, [getAccounts]);

  const loaderStyles = {
    width: '20px',
    height: '20px',
  };

  const renderTableBody = (userAccounts) => {
    if (!userAccounts.length) {
      return (
        <tbody>
        <tr>
          <td colSpan="5">No account records.</td>
        </tr>
      </tbody>
      );
    }

    return (
      <tbody>
        {userAccounts.map((account, key) => {
          const id = key + 1;
          return (
            <tr key={key}>
              <td>{id}</td>
              <td>{account.accountNo}</td>
              <td>{account.accountName}</td>
              <td>{`CAD$${Number.parseFloat(account.openingBalance).toFixed(2)}`}</td>
              <td>{`CAD$${Number.parseFloat(account.currentBalance).toFixed(2)}`}</td>
            </tr>
          );
        })}
      </tbody>
    );
  };

  const renderTable = (userAccounts) => {
    return (
      <div className="table">
        <table>
          <thead>
            <tr>
              <th>S/N</th>
              <th>AccountNo</th>
              <th>AccountName</th>
              <th>Opening Bal</th>
              <th>Current Bal</th>
            </tr>
          </thead>
          {renderTableBody(userAccounts)}
        </table>
      </div>
    );
  };

  return (
    <div>
      <Header />
      <div className="account-container">
        <div>
          {isLoading && <Loader customStyles={loaderStyles} message="fetching account records, hang tight..." />}
          {!isLoading && (<p>My Accounts</p>)}
          {!isLoading && renderTable(accounts)}
        </div>
      </div>
    </div>
  );
};

HomePage.propTypes = {
  getAccounts: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  userErrorMessage: PropTypes.string.isRequired,
  accounts: PropTypes.array.isRequired,
};

const mapStateToProps = ({ account }) => ({
  isLoading: account.isLoading,
  userErrorMessage: account.errorMessage,
  accounts: account.accounts,
});

const mapDispatchToProps = {
  getAccounts: accountActions.getAccounts,
};

export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
