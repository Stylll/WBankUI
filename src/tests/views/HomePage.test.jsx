import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import React from 'react';

import HomePage from '../../views/HomePage/index';
import store from '../../redux/store/index';
import { getAccountsSuccess } from '../../redux/actionCreators/accountActions';

const accounts = [
  {
    accountName: 'AccountA', openingBalance: 300, accountNo: '234', currentBalance: 300,
  },
  {
    accountName: 'AccountB', openingBalance: 500, accountNo: '432', currentBalance: 500,
  },
];

describe('HomePage Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(
        <Provider store={store}>
            <HomePage />
        </Provider>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render fetching account records', () => {
    const { baseElement, getByTestId, getByText } = render(
        <Provider store={store}>
            <HomePage />
        </Provider>,
    );
    expect(getByText('fetching account records, hang tight...')).toBeInTheDocument();
    store.dispatch(getAccountsSuccess([]));
    expect(getByText('No account records.')).toBeInTheDocument();
  });

  it('should render list of accounts', () => {
    const { baseElement, getByTestId, getByText } = render(
        <Provider store={store}>
            <HomePage />
        </Provider>,
    );
    store.dispatch(getAccountsSuccess(accounts));
    expect(getByText('AccountA')).toBeInTheDocument();
    expect(getByText('AccountB')).toBeInTheDocument();
  });

  it('should render and close create account form', () => {
    const { getByTestId, getByText, queryByText } = render(
        <Provider store={store}>
            <HomePage />
        </Provider>,
    );
    fireEvent.click(getByTestId('create-account'));
    expect(getByText('Create your bank account')).toBeInTheDocument();
    fireEvent.click(getByTestId('modal-close'));
    expect(queryByText('Create your bank account')).not.toBeInTheDocument();
  });

  it('should render and close create deposit form', () => {
    const {
      baseElement, getByTestId, getByText, queryByText,
    } = render(
        <Provider store={store}>
            <HomePage />
        </Provider>,
    );
    fireEvent.click(getByTestId('create-deposit'));
    expect(getByText('Make a deposit')).toBeInTheDocument();
    fireEvent.click(getByTestId('modal-close'));
    expect(queryByText('Make a deposit')).not.toBeInTheDocument();
  });

  it('should render and close create withdrawal form', () => {
    const {
      getByTestId, getByText, queryByText,
    } = render(
        <Provider store={store}>
            <HomePage />
        </Provider>,
    );
    fireEvent.click(getByTestId('create-withdrawal'));
    expect(getByText('Make a withdrawal')).toBeInTheDocument();
    fireEvent.click(getByTestId('modal-close'));
    expect(queryByText('Make a withdrawal')).not.toBeInTheDocument();
  });

  it('should render and close create transfer form', () => {
    const { queryByText, getByTestId, getByText } = render(
        <Provider store={store}>
            <HomePage />
        </Provider>,
    );
    fireEvent.click(getByTestId('create-transfer'));
    expect(getByText('Make a transfer')).toBeInTheDocument();
    fireEvent.click(getByTestId('modal-close'));
    expect(queryByText('Make a transfer')).not.toBeInTheDocument();
  });
});
