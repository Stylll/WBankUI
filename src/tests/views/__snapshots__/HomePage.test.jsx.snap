// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Test should render properly 1`] = `
<body>
  <div>
    <div>
      <div
        class="header container"
      >
        <div
          class="link"
        >
          WBank
        </div>
        <div
          class="topRight"
        />
      </div>
      <div
        class="bottom container"
      />
      <div
        class="navigation-container"
      >
        <div
          class="navbox create"
          data-testid="create-account"
        >
          Create Account
        </div>
        <div
          class="navbox deposit"
          data-testid="create-deposit"
        >
          Deposit
        </div>
        <div
          class="navbox withdrawal"
          data-testid="create-withdrawal"
        >
          Withdrawal
        </div>
        <div
          class="navbox transfer"
          data-testid="create-transfer"
        >
          Transfer
        </div>
      </div>
      <div
        class="account-container"
      >
        <div>
          <div
            class="loader-container "
          >
            <div
              class="loader"
              style="width: 20px; height: 20px;"
            />
            <span
              class="message"
            >
              fetching account records, hang tight...
            </span>
          </div>
        </div>
      </div>
      <div
        class="modal"
      />
      <div
        class="modal"
      />
      <div
        class="modal"
      />
      <div
        class="modal"
      />
    </div>
  </div>
</body>
`;
